File: tombalator-ui/src/utils/cardGenerator.ts
import type { TombalaCard } from '../components/CardSelection'

// Card color themes
const CARD_THEMES = [
  'theme-red',
  'theme-blue',
  'theme-green',
  'theme-purple',
  'theme-orange',
  'theme-teal',
  'theme-pink',
  'theme-yellow',
]

// Generate a random number in a specific range
function getRandomInRange(min: number, max: number, exclude: Set<number>): number {
  const available = Array.from({ length: max - min + 1 }, (_, i) => min + i).filter(
    (n) => !exclude.has(n)
  )
  return available[Math.floor(Math.random() * available.length)]
}

// Generate a single Tombala card
export function generateTombalaCard(): TombalaCard {
  const cardId = `card_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`
  const rows: (number | null)[][] = []
  const usedNumbers = new Set<number>()
  const theme = CARD_THEMES[Math.floor(Math.random() * CARD_THEMES.length)]

  // Column ranges for Tombala (1-9, 10-19, 20-29, ..., 80-90)
  const columnRanges = [
    { min: 1, max: 9 },
    { min: 10, max: 19 },
    { min: 20, max: 29 },
    { min: 30, max: 39 },
    { min: 40, max: 49 },
    { min: 50, max: 59 },
    { min: 60, max: 69 },
    { min: 70, max: 79 },
    { min: 80, max: 90 },
  ]

  // Generate 3 rows
  for (let rowIndex = 0; rowIndex < 3; rowIndex++) {
    const row: (number | null)[] = []
    const rowUsedNumbers = new Set<number>()

    // Each row has 5 numbers and 4 empty cells
    const numberPositions = [0, 1, 2, 3, 4, 5, 6, 7, 8]
      .sort(() => Math.random() - 0.5)
      .slice(0, 5)
      .sort((a, b) => a - b)

    for (let colIndex = 0; colIndex < 9; colIndex++) {
      if (numberPositions.includes(colIndex)) {
        // Place a number from the column's range
        const range = columnRanges[colIndex]
        const number = getRandomInRange(range.min, range.max, rowUsedNumbers)
        row.push(number)
        rowUsedNumbers.add(number)
        usedNumbers.add(number)
      } else {
        // Empty cell
        row.push(null)
      }
    }

    rows.push(row)
  }

  return {
    id: cardId,
    rows,
    theme,
  }
}

// Generate multiple cards
export function generateCards(count: number): TombalaCard[] {
  return Array.from({ length: count }, () => generateTombalaCard())
}

File: server/src/main/kotlin/game/CardGenerator.kt
package com.tombalator.game

import kotlin.random.Random

object CardGenerator {
    // Card color themes
    private val CARD_THEMES = listOf(
        "theme-red",
        "theme-blue",
        "theme-green",
        "theme-purple",
        "theme-orange",
        "theme-teal",
        "theme-pink",
        "theme-yellow"
    )

    // Column ranges for Tombala (1-9, 10-19, 20-29, ..., 80-90)
    private val COLUMN_RANGES = listOf(
        1 to 9,
        10 to 19,
        20 to 29,
        30 to 39,
        40 to 49,
        50 to 59,
        60 to 69,
        70 to 79,
        80 to 90
    )

    /**
     * Generates a single Tombala card following the rules:
     * - 15 numbers total (5 per row × 3 rows)
     * - Each column (decade) can have max 3 numbers across all rows
     * - Each column must have at least 1 number
     */
    fun generateCard(): TombalaCard {
        val cardId = "card_${System.currentTimeMillis()}_${Random.nextLong(0, Long.MAX_VALUE).toString(36)}"
        val theme = CARD_THEMES.random()
        
        // Initialize 3 rows with 9 null cells each
        val rows = mutableListOf<MutableList<Int?>>()
        for (i in 0 until 3) {
            rows.add(MutableList(9) { null })
        }
        
        // Track how many numbers are used in each column (max 3 per column)
        val columnCounts = IntArray(9) { 0 }
        val allUsedNumbers = mutableSetOf<Int>()
        
        // Step 1: Ensure each column has at least 1 number
        // Assign one number to each column first
        for (colIndex in 0 until 9) {
            val range = COLUMN_RANGES[colIndex]
            val availableNumbers = (range.first..range.second).filter { it !in allUsedNumbers }
            
            if (availableNumbers.isNotEmpty()) {
                // Randomly select a row for this column
                val rowIndex = Random.nextInt(0, 3)
                val number = availableNumbers.random()
                
                rows[rowIndex][colIndex] = number
                allUsedNumbers.add(number)
                columnCounts[colIndex]++
            }
        }
        
        // Step 2: Fill remaining positions to get 5 numbers per row
        // Each row needs 5 numbers total, and we've already placed some
        for (rowIndex in 0 until 3) {
            val currentNumbersInRow = rows[rowIndex].count { it != null }
            val numbersNeeded = 5 - currentNumbersInRow
            
            if (numbersNeeded > 0) {
                // Find empty positions in this row
                val emptyPositions = rows[rowIndex].mapIndexedNotNull { index, value ->
                    if (value == null) index else null
                }.shuffled()
                
                // Try to fill empty positions
                var numbersAdded = 0
                for (colIndex in emptyPositions) {
                    if (numbersAdded >= numbersNeeded) break
                    
                    // Check if we can add a number to this column (max 3 per column)
                    if (columnCounts[colIndex] < 3) {
                        val range = COLUMN_RANGES[colIndex]
                        val availableNumbers = (range.first..range.second)
                            .filter { it !in allUsedNumbers }
                        
                        if (availableNumbers.isNotEmpty()) {
                            val number = availableNumbers.random()
                            rows[rowIndex][colIndex] = number
                            allUsedNumbers.add(number)
                            columnCounts[colIndex]++
                            numbersAdded++
                        }
                    }
                }
            }
        }
        
        // Convert to immutable lists
        val finalRows = rows.map { it.toList() }
        
        return TombalaCard(
            id = cardId,
            rows = finalRows,
            theme = theme
        )
    }

    /**
     * Generates multiple cards
     */
    fun generateCards(count: Int): List<TombalaCard> {
        return (0 until count).map { generateCard() }
    }
}

